google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/AppEngineDeploymentConfigurationPanel.java

```diff
@@ -17,12 +17,25 @@
 package com.google.cloud.tools.intellij.appengine.cloud;
 
 import com.google.cloud.tools.intellij.resources.ProjectSelector;
+import com.google.cloud.tools.intellij.ui.BrowserOpeningHyperLinkListener;
+import com.google.cloud.tools.intellij.util.GctBundle;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Strings;
 
+import com.intellij.remoteServer.configuration.deployment.DeploymentSource;
+import com.intellij.ui.HyperlinkLabel;
+import com.intellij.ui.components.JBTextField;
+import com.intellij.util.ui.tree.TreeModelAdapter;
+
+import org.jetbrains.annotations.NotNull;
+
+import javax.swing.JCheckBox;
 import javax.swing.JLabel;
 import javax.swing.JPanel;
+import javax.swing.event.TreeModelEvent;
 
 /**
- * Common App Engine deployment configuration UI shared by flexible and standard.
+ * Common App Engine deployment configuration UI shared by flexible and standard deployments.
  */
 public class AppEngineDeploymentConfigurationPanel {
```

Comment:
Correct, the recommendation is to follow Effective Java's advice: "design and document for inheritance or else prohibit it." So for classes, the easiest rule to follow is always make them final unless it should be sub-classed, in which case its usage should be documented.

For instance variables, final should be used if the field should not be reassigned. This is the preferred behavior for most fields too to prevent mutability issues for the field itself.

For method params, it's mostly discouraged because it causes more harm than good (burden of enforcement, extra noise, etc.). Also as a quick note, as of Java 8, it's no longer necessary to use final on local variables to make them accessible from an inner class, which is often why local vars are final.